---
title: "DMRs identification with mCSEA package"
author:
- name: Jordi Martorell-Marugán
  affiliation:
  - Bioinformatics Unit. GENYO, Centre for Genomics and Oncological Research
- name: Pedro Carmona-Sáez
  affiliation:
  - Bioinformatics Unit. GENYO, Centre for Genomics and Oncological Research
  email: pedro.carmona@genyo.es
package: mCSEA
date: "`r doc_date()`"
abstract: >
  mCSEA (methylathed CpGs Set Enrichment Analysis) searches Differentially
  Methylated Regions (DMRs) between conditions using methylation data from 
  Illumina's 450k or EPIC microarrays. The evaluated DMRs are predefined 
  regions (promoters, gene bodies, CpG Islands and user-defined regions).
  This package contains functions to rank the CpG probes, to apply a 
  GSEA analysis for DMRs identification, and to plot the results.
output:
  BiocStyle::pdf_document
vignette: >
  %\VignetteIndexEntry{Predefined DMRs identification with mCSEA package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
references:
- id: du10
  title: "Comparison of Beta-value and M-value methods for quantifying 
methylation levels by microarray analysis"
  author:
  - family: Du
    given: P
  - family: Zhang
    given: X
  - family: Huang
    given: C
  - family: Jafari
    given: N
  - family: Kibbe
    given: WA
  - family: Hou
    given: L
  - family: Lin
    given: SM
  container-title: BMC Bioinformatics
  issued:
    year: 2010
  volume: 11
  issue: 587
---

# Previous steps

This package is not intended to process raw methylation data. The input of 
mCSEA is tipically a matrix with the processed beta-values for each probe 
and sample. If you start from the raw methylation files (like .idat files) you 
should first preprocess the data with any of the available packages for 
that purpose (e.g. `r Biocpkg("minfi")` or `r Biocpkg("ChAMP")`). Minfi 
includes functions to get a matrix of beta-values (_getBeta()_) or M-values 
(_getM()_). ChAMP output class depends on the type of analysis performed. For 
instance, _champ.norm()_ function returns a matrix, while _champ.load()_ 
returns a list of results, and one of them is a beta-values matrix. So mCSEA 
is totally compatible with minfi and ChAMP outputs as long as a matrix with 
the methylation values is obtained.

# Step 1: Ranking CpGs probes

To run a mCSEA analysis, you must rank all the evaluated CpGs probes with some 
metric (e.g. t-statistic, Fold-Change...). You can use _rankProbes()_ function 
for that aim, or prepare a ranked list with the same structure as the 
_rankProbes()_ output.

We load sample data to show how _rankProbes()_ works: 

```{r}
library(mCSEA)
data(mcseadata)
```

We loaded to our R environment **betaTest** and **phenoTest** objects, in 
addition to association objects (we will talk about these after). **betaTest** 
is a matrix with the beta-values of 10000 EPIC probes for 20 samples. 
**phenoTest** is a dataframe with the explanatory variable and covariates 
associated to the samples. When you load your own data, the structure of 
your objects should be similar.

```{r}
head(betaTest, 3)
print(phenoTest)
```

_rankProbes()_ function uses these two objects as input and apply a linear 
model with `r Biocpkg("limma")` package. By default, _rankProbes()_ 
considers the first column of the phenotypes table as the explanatory variable 
in the model (e.g. cases and controls) and does not take into account any 
covariate to adjust the models. You can change this behaviour modifying 
**explanatory** and **covariates** options.

By default, _rankProbes()_ assumes that the methylation data object contains 
beta-values and transform them to M-values before calculating the linear 
models. If your methylation data object contains M-values, you must specify 
it (typeInput = "M"). You can also use beta-values for models calculation 
(typeAnalysis = "beta"), although we do not recommend it due to it has been 
proven that M-values better accomplish the statistical assumptions of limma 
analysis (@du10).

```{r}
myRank <- rankProbes(betaTest, phenoTest, refGroup = "Control")
```
**myRank** is a named vector with the t-values for each CpG probe.

```{r}
head(myRank)
```

You can also supply _rankProbes()_ function with a SummarizedExperiment 
object. In that case, if you don't specify a **pheno** object, phenotypes 
will be extracted from the SummarizedExperiment object with _colData()_ 
function.

# Step 2: Searching DMRs in predefined regions

Once you calculated a score for each CpG, you can perform the mCSEA analysis. 
For that purpose, you should use _mCSEATest()_ function. This function takes 
as input the vector generated in the previous step. By default, it searches 
for differentially methylated promoters, gene bodies and CpG Islands. You can 
specify the regions you want to test with _regionsTypes_ option. _minCpGs_ 
option specifies the minimum amount of CpGs in a region to be considered in 
the analysis (5 by default). You can increase the number of processors to use 
with _nproc_ option (recommended if you have enough computational resources). 
By default, mCSEA performs 10000 
permutations to calculate p-values, but you can change that with _nperm_ 
option. Finally, you should specify if the array platform is 450k or EPIC with 
the _platform_ option. 
Note that _mCSEATest()_ performs permutations to get the p-values, so each 
time it is executed, the results are not exactly the same. To avoid that, we 
recommend to include a _set.seed()_ in order to get reproducible results. 

```{r}
set.seed(123)
myResults <- mCSEATest(myRank, regionsTypes = "promoters", platform = "EPIC")
```

_mCSEATest()_ returns a list with the GSEA results and the association objects 
for each region type analyzed.

```{r}
ls(myResults)
```

_promoters_ is a data frame with the following columns (partially extracted 
from `r Biocpkg("fgsea")` help):

* *pval:* Estimated P-value.
* *padj:* P-value adjusted by BH method.
* *ES:* Enrichment score.
* *NES:* Normalized enrichment score by number of CpGs associated to the 
feature.
* *nMoreExtreme:* Number of times a random gene set had a more extreme 
enrichment score value.
* *size:* Number of CpGs associated to the feature.
* *leadingEdge:* Leading edge CpGs which drive the enrichment.

```{r}
head(myResults[["promoters"]][,-7])
```

On the other hand, _promoters_association_ is a list with the CpG probes 
associated to each feature:

```{r}
head(myResults[["promoters_association"]], 3)
```

In addition, you can also provide a custom association object between CpG 
probes and regions (_customAnnotation_ option). This object should be a list 
with a structure similar to this:

```{r}
head(assocGenes450k, 3)
```

# Step 3: Plotting the results

Once you found some DMRs, you can make a plot with the genomic context of the 
interesting ones. For that, you must provide _mCSEAPlot()_ function with the 
methylation matrix, the phenotype object and the _mCSEATest()_ results, and 
you must specify which type of region you want to plot and the name of the 
plotted region (e.g. gene name). There are some graphical parameters you can 
adjust (see _mCSEAPlot()_ help). Take into account that this function connects 
to some online servers in order to get genomic information. For that reason, 
this function could take some minutes to finish the plot, specially the first 
time it is executed.

```{r, message = FALSE, results='hide'}
mCSEAPlot(betaTest, phenoTest, myResults, regionType = "promoters", 
           region = "CLIC6", platform="EPIC",
           transcriptAnnotation = "symbol", makePDF = FALSE)
```

You can also plot the GSEA results for a region with _mCSEAPlotGSEA()_ 
function.

```{r}
mCSEAPlotGSEA(myRank, myResults, regionType = "promoters", region = "CLIC6")
```

# Session info
```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References
