#' Plot mCSEA results
#'
#' Generate a graphic with the genomic context of the selected region, showing
#' methylation status at each CpG site of different samples groups
#'
#' @param data A data frame or a matrix containing Illumina's CpG probes in rows
#'  and samples in columns
#' @param pheno A data frame or a matrix containing samples in rows and
#' covariates in columns
#' @param mCSEAResults The object generated by mCSEATest function
#' @param regionType The region type to be represented. Must be one of
#' "promoters", "genes", "CGI" or "custom"
#' @param region The region of interest to be represented (e.g. gene name, CGI
#' name...)
#' @param column The column name or number from pheno used to split the samples
#' into groups (first column is used by default)
#' @param platform Platform used to get the methylation data (450k or EPIC)
#' @param extend The number of base pairs to extend the plot around the region
#' of interest (default = 1000 bp)
#' @param chromosome If TRUE, represent the chromosome and genome axis
#' @param leadingEdge If TRUE, represent the bars indicating if each CpG belongs
#'  to the mCSEA leading edge (green) or not (red)
#' @param CGI If TRUE, represent the annotated CpG islands
#' @param genes If TRUE, represent the annotated genes
#' @param transcriptAnnotation Labels showed at the genes track. Must be one of
#' "transcript" (default), "symbol", "gene", "exon" or "feature"
#' @param makePDF If TRUE, generate save the plot in pdf format in the working
#' directory. Otherwise, draw the plot in the active graphics window
#'
#' @return 'NULL'
#'
#' @author Jordi Martorell Marug√°n, \email{jordi.martorell@@genyo.es}
#'
#' @seealso \code{\link{rankProbes}}, \code{\link{mCSEATest}},
#' \code{\link{mCSEAPlotGSEA}}
#'
#' @examples
#' library(mCSEAdata)
#' data(mcseadata)
#' \dontrun{
#' myRank <- rankProbes(betaTest, phenoTest, refGroup = "Control")
#' set.seed(123)
#' myResults <- mCSEATest(myRank, regionsTypes = "promoters",
#' platform = "EPIC")
#' }
#' data(precomputedmCSEA)
#' mCSEAPlot(betaTest, phenoTest, myResults, "promoters", "CLIC6",
#' platform = "EPIC", transcriptAnnotation = "symbol", makePDF = FALSE)
#' @export


mCSEAPlot <- function(data, pheno, mCSEAResults, regionType, region,
                        column = 1, platform = "450k", extend = 1000,
                        chromosome = TRUE, leadingEdge = TRUE, CGI = FALSE,
                        genes = TRUE, transcriptAnnotation = "transcript",
                        makePDF = TRUE){

    progress <- utils::txtProgressBar(min=0, max=10, style=3)

    # Check input objects
    if (!any(class(data) == "data.frame" | class(data) == "matrix")){
        stop("data must be a data frame or a matrix")
    }

    if (!any(class(pheno) == "data.frame" | class(pheno) == "matrix")){
        stop("pheno must be a data frame or a matrix")
    }

    if (any(class(mCSEAResults) != "list" | length(mCSEAResults) < 2)){
        stop("mCSEAResults must be the object generated
                by mCSEATest function")
    }

    if (class(region) != "character"){
        stop("region must be a character object")
    }

    if (!any(class(column) != "character" | class(column) != "numeric")){
        stop("column must be a character or numeric object")
    }

    if (class(platform) != "character"){
        stop("typeInput must be a character object")
    }

    if (any(class(extend) != "numeric" | extend < 0)){
        stop("extend must be a positive number")
    }

    if (class(transcriptAnnotation) != "character"){
        stop("transcriptAnnotation must be a character object")
    }

    if (class(chromosome) != "logical"){
        stop("chromosome must be a logical object (TRUE7FALSE)")
    }

    if (class(leadingEdge) != "logical"){
        stop("leadingEdge must be a logical object (TRUE7FALSE)")
    }

    if (class(CGI) != "logical"){
        stop("CGI must be a logical object (TRUE7FALSE)")
    }

    if (class(genes) != "logical"){
        stop("genes must be a logical object (TRUE7FALSE)")
    }

    if (class(makePDF) != "logical"){
        stop("makePDF must be a logical object (TRUE7FALSE)")
    }

    transcriptAnnotation <- match.arg(transcriptAnnotation,
                                        choices=c("transcript", "symbol",
                                                "gene", "exon", "feature"))



    # Get the appropiate association

    regionType <- match.arg(regionType, choices=c("promoters", "genes", "CGI",
                                                    "custom"))

    assoc <- mCSEAResults[[paste(regionType, "association", sep="_")]]

    leadingCpGs <- strsplit(mCSEAResults[[regionType]]
                            [region, "leadingEdge"],", ")[[1]]

    cgs.plot <- assoc[[region]]

    CpGs <- data[cgs.plot,]
    Phenotype <- pheno[,column]

    utils::setTxtProgressBar(progress, 1)

    if (platform == "450k") {
        rsobj <- minfi::RatioSet(CpGs, annotation=c(
            array="IlluminaHumanMethylation450k", annotation="ilmn12.hg19"))
        utils::setTxtProgressBar(progress, 2)
        annot <- minfi::getAnnotation(rsobj, what = c("Locations", "Other"))
    }
    else {
        rsobj <- minfi::RatioSet(CpGs, annotation=c(
            array="IlluminaHumanMethylationEPIC",annotation="ilm10b2.hg19"))
        utils::setTxtProgressBar(progress, 2)
        annot <- minfi::getAnnotation(rsobj, what = c("Locations", "Other"))
    }

    utils::setTxtProgressBar(progress, 3)

    positions <- annot[c(cgs.plot), "pos"]
    Chromosome <- annot[c(cgs.plot[1]), "chr"]
    From <- min(positions) - extend
    To <- max(positions) + extend

    #Check for annotation
    annotSubset <- annot[rownames(annot) %in% cgs.plot,]
    probes <- rownames(annotSubset)

    Phenotype <- factor(as.character(Phenotype))


    # Subset data
    dataSubset <- data[na.omit(match(probes, rownames(data))),]

    utils::setTxtProgressBar(progress, 4)

    if (chromosome) {
        #Ideogram
        itrack <- Gviz::IdeogramTrack(genome="hg19", chromosome=Chromosome,
                                    bands=mCSEA::bandTable)

        #Genome axis
        gtrack <- Gviz::GenomeAxisTrack()
    }
    utils::setTxtProgressBar(progress, 5)


    if (genes){
        #Genes track from UCSC
        biomTrack <- Gviz::BiomartGeneRegionTrack(genome="hg19",
                                                chromosome=Chromosome,
                                                start=From, end=To ,
                                                name="ENSEMBL annotation",
                                                stacking="squish",
                                                just.group="above",
                                                transcriptAnnotation=
                                                    transcriptAnnotation)
    }
    utils::setTxtProgressBar(progress, 6)


    #CpG islands
    if(CGI) {
        cpgIslands <- Gviz::UcscTrack(genome="hg19", chromosome=Chromosome,
                                    track="cpgIslandExt", from=From,
                                    to=To,
                                    trackType="AnnotationTrack",
                                    start="chromStart", end="chromEnd",
                                    showId=FALSE, shape="box",
                                    fill="#006400", name="CpG Islands")
    }
    utils::setTxtProgressBar(progress, 7)

    #Data track
    annotSubset <- annotSubset[match(rownames(dataSubset),
                                    rownames(annotSubset)),]

    #Reorder columns
    dataSubsetOrdered <- dataSubset[,order(Phenotype)]
    phenotypeOrdered <- Phenotype[order(Phenotype)]

    nSamples <- nrow(dataSubsetOrdered)

    #Place data
    dataValues <- data.frame(c(data.frame(annotSubset[,c("pos","pos","chr")]),
                            as.data.frame(dataSubsetOrdered)))
    rownames(dataValues) <- rownames(dataSubsetOrdered)
    colnames(dataValues)[1] <- "start"
    colnames(dataValues)[2] <- "end"
    colnames(dataValues)[3] <- "chromosome"
    dataValues[2] <- dataValues[2] + 1
    dtrack <- Gviz::DataTrack(dataValues, genome="hg19",
                            type=c("g","p","a"),
                            name="DNA Methylation",
                            groups=phenotypeOrdered)
    utils::setTxtProgressBar(progress, 8)


    if (leadingEdge){

        leadingValues <- data.frame(c(dataValues[,1:3]), -1,
                                    row.names=rownames(dataValues))

        for (site in rownames(leadingValues)) {
            if (site %in% leadingCpGs){
                leadingValues[site,4] <- 1
            }
        }

        if (sum(leadingValues[,4]) == nrow(leadingValues)) {
            dtrack2 <- Gviz::DataTrack(leadingValues, genome="hg19",
                                    type="heatmap",
                                    name="KS leading edge",
                                    ncolor=2, gradient=c("green", "red"),
                                    showAxis=FALSE)
        }

        else {
            dtrack2 <- Gviz::DataTrack(leadingValues, genome="hg19",
                                    type="heatmap",
                                    name="KS leading edge",
                                    ncolor=2, gradient=c("red", "green"),
                                    showAxis=FALSE)
        }
    }
    utils::setTxtProgressBar(progress, 9)



    #Scale PDF

    if (genes){
        lengthGenes <- length(unique(GenomicRanges::restrict(attr(
            biomTrack, "range"), From, To)$transcript))
        heightPdf <- sum(chromosome, CGI, leadingEdge) + 0.5*lengthGenes + 1
        heightGeneTrack <- 2 + 0.5*lengthGenes
    }

    else {
        heightPdf <- sum(chromosome, CGI, leadingEdge, genes) + 2
    }

    if (makePDF){
        pdf(paste0(region,".pdf"), height=heightPdf)
    }

    # Plot tracks

    if (chromosome) {
        if (leadingEdge & CGI & genes){
            Gviz::plotTracks(list(itrack, gtrack, dtrack, dtrack2, cpgIslands,
                                biomTrack),
                            from=From, to=To,
                            background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(0.5,1,3,1,2,heightGeneTrack),
                            background.panel="white",
                            background.title="#2e2e2e")
        }
        else if (CGI & genes){
            Gviz::plotTracks(list(itrack, gtrack, dtrack,
                                cpgIslands, biomTrack),
                            from=From, to=To, background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(0.5,1,3,2,heightGeneTrack),
                            background.panel="white",
                            background.title="#2e2e2e")
        }
        else if (leadingEdge & genes){
            Gviz::plotTracks(list(itrack, gtrack, dtrack, dtrack2, biomTrack),
                            from=From, to=To, background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(0.5,1,3,1,heightGeneTrack),
                            background.panel="white",
                            background.title="#2e2e2e")
        }
        else if (leadingEdge & CGI){
            Gviz::plotTracks(list(itrack, gtrack, dtrack, dtrack2, cpgIslands),
                            from=From, to=To, background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(0.5,1,3,1,2), background.panel="white",
                            background.title="#2e2e2e")
        }
        else if (leadingEdge){
            Gviz::plotTracks(list(itrack, gtrack, dtrack, dtrack2), from=From,
                            to=To, background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(0.5,1,3,1), background.panel="white",
                            background.title="#2e2e2e")
        }
        else if (CGI){
            Gviz::plotTracks(list(itrack, gtrack, dtrack, cpgIslands),
                            from=From,
                            to=To, background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(0.5,1,3,2), background.panel="white",
                            background.title="#2e2e2e")
        }
        else if (genes){
            Gviz::plotTracks(list(itrack, gtrack, dtrack, biomTrack),
                            from=From,
                            to=To, background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(0.5,1,3,heightGeneTrack),
                            background.panel="white",
                            background.title="#2e2e2e")
        }
        else {
            Gviz::plotTracks(list(itrack, gtrack, dtrack), from=From, to=To,
                            background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(0.5,1,3), background.panel="white",
                            background.title="#2e2e2e")
        }
    }

    else if (leadingEdge) {
        if (CGI & genes){
            Gviz::plotTracks(list(dtrack, dtrack2, cpgIslands, biomTrack),
                            from=From, to=To, background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(3,1,2,heightGeneTrack),
                            background.panel="white",
                            background.title="#2e2e2e")
        }
        else if (genes){
            Gviz::plotTracks(list(dtrack, dtrack2, biomTrack), from=From,
                            to=To,
                            background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(3,1,heightGeneTrack),
                            background.panel="white",
                            background.title="#2e2e2e")
        }
        else if (CGI){
            Gviz::plotTracks(list(dtrack, dtrack2, cpgIslands), from=From,
                            to=To,
                            background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(3,1,2), background.panel="white",
                            background.title="#2e2e2e")
        }
        else {
            Gviz::plotTracks(list(dtrack, dtrack2), from=From, to=To,
                            background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(3,1), background.panel="white",
                            background.title="#2e2e2e")
        }
    }

    else if (CGI) {
        if (genes){
            Gviz::plotTracks(list(dtrack, cpgIslands, biomTrack), from=From,
                            to=To, background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(3,2,heightGeneTrack),
                            background.panel="white",
                            background.title="#2e2e2e")
        }
        else {
            Gviz::plotTracks(list(dtrack, cpgIslands), from=From, to=To,
                            background.panel="#FFFFFF",
                            background.title="darkblue", lwd=1,
                            legend=TRUE,
                            sizes=c(3,2), background.panel="white",
                            background.title="#2e2e2e")
        }
    }

    else if (genes) {
        Gviz::plotTracks(list(dtrack, biomTrack), from=From, to=To,
                        background.panel="#FFFFFF",
                        background.title="darkblue", lwd=1, legend=TRUE,
                        sizes=c(3,heightGeneTrack),
                        background.panel="white",
                        background.title="#2e2e2e")
    }

    else {
        Gviz::plotTracks(list(dtrack), from=From, to=To,
                        background.panel="#FFFFFF",
                        background.title="darkblue", lwd=1, legend=TRUE,
                        sizes=c(3), background.panel="white",
                        background.title="#2e2e2e")
    }

    if (makePDF){
        dev.off()
    }

    utils::setTxtProgressBar(progress, 10)

}
